v3/mathjax3-ts/core/MathItem.ts

```diff
@@ -0,0 +1,331 @@
+    public bbox: BBox = {};
+    public inputData: OptionList = {};
+    public outputData: OptionList = {};
+
+    /*
+     * @param{string} math      The math expression for this item
+     * @param{Inputjax} jax     The input jax to use for this item
+     * @param{boolean} display  True if display mode, false if inline
+     * @param{Location} start   The starting position of the math in the document
+     * @param{Location} end     The ending position of the math in the document
+     * @constructor
+     */
+    constructor (math: string, jax: InputJax, display: boolean = true,
+                 start: Location = {i: 0, n: 0, delim: ''},
+                 end: Location = {i: 0, n: 0, delim: ''}) {
+        this.math = math;
+        this.inputJax = jax;
+        this.display = display;
+        this.start = start;
+        this.end = end;
+        this.root = null;
+        this.typesetRoot = null;
+        this.metrics = {} as Metrics;
+        this.bbox = {};
+        this.inputData = {};
+        this.outputData = {};
+    }
+
+    /*
+     * @override
+     */
+    public compile(document: MathDocument) {
+        if (this.state() < STATE.COMPILED) {
+            this.root = this.inputJax.compile(this);
+            this.state(STATE.COMPILED);
+        }
+    }
+
+    /*
+     * @override
+     */
+    public typeset(document: MathDocument) {
+        if (this.state() < STATE.TYPESET) {
+            if (this.display === null) {
+                this.typesetRoot = document.outputJax.escaped(this, document);
+            } else {
+                this.typesetRoot = document.outputJax.typeset(this, document);
+            }
+            this.state(STATE.TYPESET);
+        }
+    }
+
+    /*
+     * @override
+     */
+    public addEventHandlers() {}
+
+    /*
+     * @override
+     */
+    public updateDocument(document: MathDocument) {}
+
+    /*
+     * @override
+     */
+    public removeFromDocument(restore: boolean = false) {}
+
+    /*
+     * @override
+     */
+    public setMetrics(em: number, ex: number, cwidth: number, lwidth: number, scale: number) {
+        this.metrics = {
+            em: em, ex: ex,
+            containerWidth: cwidth,
+            lineWidth: lwidth,
+            scale: scale
+        };
+    }
+
+    /*
+     * @override
+     */
+    public state(state: number = null, restore: boolean = false) {
+        if (state != null) {
+            if (state < STATE.INSERTED && this._state >= STATE.INSERTED) {
+                this.removeFromDocument(restore);
+            }
+            if (state < STATE.TYPESET && this._state >= STATE.TYPESET) {
+                this.bbox = {};
+                this.outputData = {};
+            }
+            if (state < STATE.COMPILED && this._state >= STATE.COMPILED) {
+                this.inputData = {};
+            }
+            this._state = state;
+        }
+        return this._state;
+    }
+
+}
+
+let STATE = AbstractMathItem.STATE;
```

Comment:
Right, different document classes can have their own states.  For example, a Markdown document might have states for whether the item has been encoded in the document (i.e., replaced by a token that identifies the math but has no Markdown commands in it, like source TeX can) or not.  I don't know what all those states may be, so can't have just a fixed set.  I suspect the ones here are sufficient, but don't know for sure.